src/common/scrub_types.cc

```diff
@@ -55,11 +55,8 @@ static void encode(const osd_shard_t& shard, bufferlist& bl) {
 
 void shard_info_wrapper::set_object(const ScrubMap::object& object)
 {
-  for (auto attr : object.attrs) {
-    bufferlist bl;
-    bl.push_back(attr.second);
-    attrs.insert(std::make_pair(attr.first, std::move(bl)));
-  }
+  // no-op, changes the comparator from std::less<void>
+  attrs = reinterpret_cast<const std::map<std::string, ceph::bufferlist>&>(object.attrs);
```

Comment:
I dislike it.

Reply:
that is naughty;  it can't be a static_cast?

Reply:
the type of `ScrubMap::object::attrs` is `std::map<std::string, ceph::buffer::ptr, std::less<>>`, while the type of `shard_info_t::attrs` is `std::map<std::string, ceph::bufferlist>`.

why do you think `ceph::buffer::ptr` can be used as `ceph::bufferlist` ? please note that the latter is an alias of `ceph::buffer::list`

Reply:
Unfortunately, `static_cast` doesn't compile. I will take a look on the transparent comparator and elaborate on changes. Thanks, Guys!

Reply:
My understanding is `shard_info_t` is a part of the public API but rather infrequently used.

Its sole client is `get_inconsistent_objects()`. The struct is used indirectly, through the `std::vector<inconsistent_obj_t>* objects` parameter.

```cpp
    /**
     * List the inconsistent objects found in a given PG by last scrub
     *
     * @param pg the placement group returned by @c pg_list()
     * @param start_after the first returned @c objects
     * @param max_return the max number of the returned @c objects
     * @param c what to do when the operation is complete and safe
     * @param objects [out] the objects where inconsistencies are found
     * @param interval [in,out] an epoch indicating current interval
     * @returns if a non-zero @c interval is specified, will return -EAGAIN i
     *          the current interval begin epoch is different.
     */
    int get_inconsistent_objects(const PlacementGroup& pg,
                                 const object_id_t &start_after,
                                 unsigned max_return,
                                 AioCompletion *c,
                                 std::vector<inconsistent_obj_t>* objects,
                                 uint32_t* interval);
```

```cpp
struct inconsistent_obj_t : obj_err_t {
...
  std::map<osd_shard_t, shard_info_t> shards;
...
};
```

If we enable the transparent comparator `shard_info_t::attrs` there is non-zero risk breaking buildability of 3rd party applications, though the risk is very small (negligible?).

Reply:
Can we use a loop to fill attrs somewhat inefficiently and worry about this later?

Reply:
Yeah, we should start from the loop; `reinterpret_cast` looks like a premature optimization.

Reply:
Switched to the loop (and left a comment).